src/backend/distributed/utils/shard_utils.c

```diff
@@ -55,3 +59,52 @@ GetLongestShardName(Oid citusTableOid, char *finalRelationName)
 
 	return longestShardName;
 }
+
+
+/*
+ * GetLongestHypotheticalShardName is a utility function that creates a hypothetical shard
+ * name for a table that is not distributed yet.
+ */
+char *
+GetLongestHypotheticalShardName(Oid parentTableOid, char *relationName)
+{
+	char *longestShardName = pstrdup(relationName);
+	CitusTableCacheEntry *cacheEntry = GetCitusTableCacheEntry(parentTableOid);
+	int shardIntervalCount = cacheEntry->shardIntervalArrayLength;
+
+	Oid savedUserId = InvalidOid;
+	int savedSecurityContext = 0;
+
+	text *sequenceName = cstring_to_text(SHARDID_SEQUENCE_NAME);
+	Oid sequenceId = ResolveRelationId(sequenceName, false);
+	Datum sequenceIdDatum = ObjectIdGetDatum(sequenceId);
+
+	GetUserIdAndSecContext(&savedUserId, &savedSecurityContext);
+	SetUserIdAndSecContext(CitusExtensionOwner(), SECURITY_LOCAL_USERID_CHANGE);
+
+	int64 newShardId = 0;
+
+	/*
+	 * pg_sequence_last_value() returns NULL if the sequence value is not yet used.
+	 * DirectFunctionCall1() gives an ERROR message on NULL return values, and that's why we
+	 * need a PG_TRY block.
+	 */
+	PG_TRY();
+	{
+		Datum lastShardIdDatum = DirectFunctionCall1(pg_sequence_last_value,
+													 sequenceIdDatum);
+		newShardId = DatumGetInt64(lastShardIdDatum) + shardIntervalCount;
+	}
+	PG_CATCH();
+	{
+		/* assume that we have a shardId with 7 digits */
+		newShardId = 9999999;
+	}
+	PG_END_TRY();
+
+	SetUserIdAndSecContext(savedUserId, savedSecurityContext);
+
+	AppendShardIdToName(&longestShardName, newShardId);
```

Comment:
should we make `newShardId + 1`?

Reply:
I don't think this is necessary. Since the value returned by `pg_sequence_last_value` will be the last used shard id value, the next consecutive shard id values should be used by this new partitions.